File: src/main/java/com/flexpoker/game/command/aggregate/DefaultGameFactory.java
package com.flexpoker.game.command.aggregate;

import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Component;

import com.flexpoker.game.command.commands.CreateGameCommand;
import com.flexpoker.game.command.events.GameCreatedEvent;
import com.flexpoker.game.command.factory.GameFactory;
import com.flexpoker.game.command.framework.GameEvent;
import com.flexpoker.game.query.dto.GameStage;

@Component
public class DefaultGameFactory implements GameFactory {

    @Override
    public Game createNew(CreateGameCommand command) {
        UUID aggregateId = UUID.randomUUID();
        return createGame(false, aggregateId, command.getGameName(),
                command.getNumberOfPlayers(),
                command.getNumberOfPlayersPerTable(),
                command.getCreatedByPlayerId(),
                command.getNumberOfMinutesBetweenBlindLevels(),
                command.getNumberOfSecondsForActionOnTimer());
    }

    @Override
    public Game createFrom(List<GameEvent> events) {
        GameCreatedEvent gameCreatedEvent = (GameCreatedEvent) events.get(0);
        Game game = createGame(true, gameCreatedEvent.getAggregateId(),
                gameCreatedEvent.getGameName(),
                gameCreatedEvent.getNumberOfPlayers(),
                gameCreatedEvent.getNumberOfPlayersPerTable(),
                gameCreatedEvent.getCreatedByPlayerId(),
                gameCreatedEvent.getNumberOfMinutesBetweenBlindLevels(),
                gameCreatedEvent.getNumberOfSecondsForActionOnTimer());
        game.applyAllHistoricalEvents(events);
        return game;
    }

    private Game createGame(boolean creatingFromEvents, UUID aggregateId,
            String gameName, int maxNumberOfPlayers,
            int numberOfPlayersPerTable, UUID createdById,
            int numberOfMinutesBetweenBlindLevels,
            int numberOfSecondsForActionOnTimer) {
        BlindSchedule blindSchedule = new BlindSchedule(numberOfMinutesBetweenBlindLevels);
        TableBalancer tableBalancer = new TableBalancer(aggregateId,
                numberOfPlayersPerTable);
        return new Game(creatingFromEvents, aggregateId, gameName,
                maxNumberOfPlayers, numberOfPlayersPerTable,
                numberOfSecondsForActionOnTimer, createdById,
                GameStage.REGISTERING, blindSchedule, tableBalancer);
    }

}

File: src/main/kotlin/com/flexpoker/game/command/aggregate/DefaultGameFactory.kt
package com.flexpoker.game.command.aggregate

import com.flexpoker.game.command.commands.CreateGameCommand
import com.flexpoker.game.command.events.GameCreatedEvent
import com.flexpoker.game.command.events.GameEvent
import com.flexpoker.game.command.factory.GameFactory
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class DefaultGameFactory : GameFactory {

    override fun createNew(command: CreateGameCommand): Game {
        return createGame(
            false, UUID.randomUUID(), command.gameName,
            command.numberOfPlayers,
            command.numberOfPlayersPerTable,
            command.createdByPlayerId,
            command.numberOfMinutesBetweenBlindLevels,
            command.numberOfSecondsForActionOnTimer
        )
    }

    override fun createFrom(events: List<GameEvent>): Game {
        val gameCreatedEvent = events[0] as GameCreatedEvent
        val game = createGame(
            true, gameCreatedEvent.aggregateId,
            gameCreatedEvent.gameName,
            gameCreatedEvent.numberOfPlayers,
            gameCreatedEvent.numberOfPlayersPerTable,
            gameCreatedEvent.createdByPlayerId,
            gameCreatedEvent.numberOfMinutesBetweenBlindLevels,
            gameCreatedEvent.numberOfSecondsForActionOnTimer
        )
        game.applyAllHistoricalEvents(events)
        return game
    }

    private fun createGame(
        creatingFromEvents: Boolean, aggregateId: UUID,
        gameName: String, maxNumberOfPlayers: Int,
        numberOfPlayersPerTable: Int, createdById: UUID,
        numberOfMinutesBetweenBlindLevels: Int,
        numberOfSecondsForActionOnTimer: Int
    ): Game {
        return Game(creatingFromEvents, aggregateId, gameName, maxNumberOfPlayers, numberOfPlayersPerTable,
            numberOfSecondsForActionOnTimer, createdById, numberOfMinutesBetweenBlindLevels)
    }
}